src/zenml/zen_stores/migrations/utils.py

```diff
@@ -0,0 +1,584 @@
+#  Copyright (c) ZenML GmbH 2022. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""ZenML database migration, backup and recovery utilities."""
+
+import json
+import os
+import shutil
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generator,
+    List,
+    Optional,
+)
+
+from pydantic import BaseModel
+from pydantic.json import pydantic_encoder
+from sqlalchemy import MetaData, func, text
+from sqlalchemy.engine import URL, Engine
+from sqlalchemy.schema import CreateTable
+from sqlmodel import (
+    create_engine,
+    select,
+)
+
+from zenml.logger import get_logger
+
+logger = get_logger(__name__)
+
+
+class MigrationUtils(BaseModel):
+    """Utilities for database migration, backup and recovery."""
+
+    url: URL
+    connect_args: Dict[str, Any]
+    engine_args: Dict[str, Any]
+
+    _engine: Optional[Engine] = None
+    _master_engine: Optional[Engine] = None
+
+    def create_engine(self, database: Optional[str] = None) -> Engine:
+        """Get the SQLAlchemy engine for a database.
+
+        Args:
+            database: The name of the database. If not set, a master engine
+                will be returned.
+
+        Returns:
+            The SQLAlchemy engine.
+        """
+        url = self.url._replace(database=database)
+        return create_engine(
+            url=url,
+            connect_args=self.connect_args,
+            **self.engine_args,
+        )
+
+    @property
+    def engine(self) -> Engine:
+        """The SQLAlchemy engine.
+
+        Returns:
+            The SQLAlchemy engine.
+        """
+        if self._engine is None:
+            self._engine = self.create_engine(database=self.url.database)
+        return self._engine
+
+    @property
+    def master_engine(self) -> Engine:
+        """The SQLAlchemy engine for the master database.
+
+        Returns:
+            The SQLAlchemy engine for the master database.
+        """
+        if self._master_engine is None:
+            self._master_engine = self.create_engine()
+        return self._master_engine
+
+    def create_drop_mysql_database(
+        self,
+        database: Optional[str] = None,
+        create: bool = True,
+        drop: bool = False,
+    ) -> None:
```

Comment:
I'm a bit confused why create and drop are sharing the same method. The two methods only really share 2 lines of code.

Reply:
I saw further down, that this method can do both in once go. I still feel like this should be two separate methods. Feel free to ignore my comment though.

Reply:
done